Tests fail vs Wagtail 4.1 with "AttributeError: 'list' object has no attribute 'bound_blocks'"

I'm hitting a similar error with my own factories, it looks like passing a list into a ListBlock is the problem.

Example error:

```
tests/test_factories.py:95 (test_custom_page_streamfield_data)
@pytest.mark.django_db
    def test_custom_page_streamfield_data():
        root_page = wagtail_factories.PageFactory(parent=None)
>       page = MyTestPageFactory(
            parent=root_page, body=[("char_array", ["bla-1", "bla-2"])]
        )

test_factories.py:99: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/base.py:40: in __call__
    return cls.create(**kwargs)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/base.py:528: in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/django.py:117: in _generate
    return super()._generate(strategy, params)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/base.py:465: in _generate
    return step.build()
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/builder.py:262: in build
    instance = self.factory_meta.instantiate(
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/factory/base.py:317: in instantiate
    return self.factory._create(model, *args, **kwargs)
../src/wagtail_factories/factories.py:66: in _create
    instance = cls._create_instance(model_class, parent, kwargs)
../src/wagtail_factories/factories.py:74: in _create_instance
    parent.add_child(instance=instance)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/treebeard/mp_tree.py:1083: in add_child
    return MP_AddChildHandler(self, **kwargs).process()
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/treebeard/mp_tree.py:387: in process
    newobj.save()
../../../.pyenv/versions/3.11.0/lib/python3.11/contextlib.py:81: in inner
    return func(*args, **kwds)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/models/__init__.py:1164: in save
    result = super().save(**kwargs)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/modelcluster/models.py:201: in save
    super().save(update_fields=real_update_fields, **kwargs)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/django/db/models/base.py:812: in save
    self.save_base(
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/django/db/models/base.py:878: in save_base
    post_save.send(
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/django/dispatch/dispatcher.py:176: in send
    return [
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/django/dispatch/dispatcher.py:177: in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/signal_handlers.py:89: in update_reference_index_on_save
    ReferenceIndex.create_or_update_for_object(instance)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/models/reference_index.py:352: in create_or_update_for_object
    references = set(cls._extract_references_from_object(object))
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/models/reference_index.py:285: in _extract_references_from_object
    yield from (
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/models/reference_index.py:285: in <genexpr>
    yield from (
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/fields.py:249: in extract_references
    yield from self.stream_block.extract_references(value)
../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/blocks/stream_block.py:334: in extract_references
    for (
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <wagtail.blocks.list_block.ListBlock object at 0x7f025b895bd0>
value = ['bla-1', 'bla-2']

    def extract_references(self, value):
>       for child in value.bound_blocks:
E       AttributeError: 'list' object has no attribute 'bound_blocks'

../../../.virtualenvs/wagtail-factories/lib/python3.11/site-packages/wagtail/blocks/list_block.py:323: AttributeError
```

--- a/src/wagtail_factories/blocks.py
+++ b/src/wagtail_factories/blocks.py
@@ -2,13 +2,13 @@ from collections import defaultdict
 
 import factory
 from factory.declarations import ParameteredAttribute
+from wagtail import VERSION as wagtail_version
 from wagtail.documents.blocks import DocumentChooserBlock
 from wagtail.images.blocks import ImageChooserBlock
 
-try:
+if wagtail_version >= (3, 0):
     from wagtail import blocks
-except ImportError:
-    # Wagtail<3.0
+else:
     from wagtail.core import blocks
 
 from wagtail_factories.builder import (
@@ -121,8 +121,6 @@ class ListBlockFactory(factory.SubFactory):
         return self.evaluate(None, None, kwargs)
 
     def evaluate(self, instance, step, extra):
-        subfactory = self.get_factory()
-
         result = defaultdict(dict)
         for key, value in extra.items():
             if key.isdigit():
@@ -132,14 +130,18 @@ class ListBlockFactory(factory.SubFactory):
                 if prefix and prefix.isdigit():
                     result[int(prefix)][label] = value
 
-        retval = []
-
+        subfactory = self.get_factory()
         force_sequence = step.sequence if self.FORCE_SEQUENCE else None
-        for index, index_params in sorted(result.items()):
-            retval.append(
-                step.recurse(subfactory, index_params, force_sequence=force_sequence)
-            )
-        return retval
+        values = [
+            step.recurse(subfactory, params, force_sequence=force_sequence)
+            for _, params in sorted(result.items())
+        ]
+
+        if wagtail_version >= (2, 16):
+            list_block_def = blocks.list_block.ListBlock(subfactory._meta.model())
+            return blocks.list_block.ListValue(list_block_def, values)
+        else:
+            return values
 
 
 class StructBlockFactory(factory.Factory):

--- a/tests/test_blocks.py
+++ b/tests/test_blocks.py
@@ -1,24 +1,40 @@
 from collections import OrderedDict
 
 import pytest
+from wagtail import VERSION as wagtail_version
 from wagtail.documents.models import Document
 from wagtail.images.models import Image
 
 import wagtail_factories
-from tests.testapp.factories import MyBlockFactory, MyTestPageWithStreamFieldFactory
+from tests.testapp.factories import (
+    MyBlockFactory,
+    MyBlockItemFactory,
+    MyTestPageWithStreamFieldFactory,
+)
 
-try:
+if wagtail_version >= (3, 0):
     from wagtail.blocks import StructValue
     from wagtail.models import Page
-except ImportError:
-    # Wagtail<3.0
+else:
     from wagtail.core.blocks import StructValue
     from wagtail.core.models import Page
 
 
+def eq_list_block_values(p, q):
+    """
+    Direct comparison of ListValue instances fails under current versions of Wagtail (<= 4.1),
+    so do a pairwise comparison of the bound blocks' values.
+
+    With Wagtail >= 2.16 we will get ListValues, prior to that just lists.
+    """
+    if wagtail_version >= (2, 16):
+        return all(map(lambda x, y: x.value == y.value, p.bound_blocks, q.bound_blocks))
+    return p == q
+
+
 @pytest.mark.django_db
 def test_list_block_factory():
-    value = MyBlockFactory(
+    computed = MyBlockFactory(
         items__0__label="label-1",
         items__0__value=1,
         items__1__label="label-2",
@@ -26,66 +42,88 @@ def test_list_block_factory():
         image__image=None,
     )
 
-    assert value == StructValue(
-        None,
-        [
-            ("title", "my title"),
-            ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
-            (
-                "items",
-                [
-                    StructValue(None, [("label", "label-1"), ("value", 1)]),
-                    StructValue(None, [("label", "label-2"), ("value", 2)]),
-                ],
-            ),
-            ("image", None),
-        ],
+    list_item_block = MyBlockItemFactory._meta.model()
+
+    expected = MyBlockFactory._meta.model().clean(
+        OrderedDict(
+            [
+                ("title", "my title"),
+                ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
+                (
+                    "items",
+                    [
+                        StructValue(
+                            list_item_block, [("label", "label-1"), ("value", 1)]
+                        ),
+                        StructValue(
+                            list_item_block, [("label", "label-2"), ("value", 2)]
+                        ),
+                    ],
+                ),
+            ]
+        )
     )
+    assert eq_list_block_values(computed["items"], expected["items"])
 
 
 @pytest.mark.django_db
 def test_block_factory():
-    value = MyBlockFactory(image__image__title="blub")
-
-    assert value == OrderedDict(
-        [
-            ("title", "my title"),
-            ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
-            ("items", []),
-            ("image", Image.objects.first()),
-        ]
+    computed = MyBlockFactory(image__image__title="blub")
+    expected = MyBlockFactory._meta.model().clean(
+        OrderedDict(
+            [
+                ("title", "my title"),
+                ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
+                ("items", []),
+                ("image", Image.objects.first()),
+            ]
+        )
     )
-
-    assert value["image"].title == "blub"
+    computed_list_value = computed.pop("items")
+    expected_list_value = expected.pop("items")
+    assert eq_list_block_values(computed_list_value, expected_list_value)
+    assert computed == expected
+    assert computed["image"].title == "blub"
 
 
 def test_block_factory_build():
-    value = MyBlockFactory.build(image__image__title="blub")
+    computed = MyBlockFactory.build(image__image__title="blub")
 
-    image = value.pop("image")
+    image = computed.pop("image")
     assert image.title == "blub"
 
-    assert value == OrderedDict(
-        [
-            ("title", "my title"),
-            ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
-            ("items", []),
-        ]
+    expected = MyBlockFactory._meta.model().clean(
+        OrderedDict(
+            [
+                ("title", "my title"),
+                ("item", OrderedDict([("label", "my-label"), ("value", 100)])),
+                ("items", []),
+            ]
+        )
     )
+
+    computed_list_value = computed.pop("items")
+    expected_list_value = expected.pop("items")
+    assert eq_list_block_values(computed_list_value, expected_list_value)
+    assert computed == expected
 
 
 @pytest.mark.django_db
 def test_block_factory_subkwarg():
-    value = MyBlockFactory(item__label="my-label", item__value=20, image__image=None)
-
-    assert value == OrderedDict(
-        [
-            ("title", "my title"),
-            ("item", OrderedDict([("label", "my-label"), ("value", 20)])),
-            ("items", []),
-            ("image", None),
-        ]
+    computed = MyBlockFactory(item__label="my-label", item__value=20, image__image=None)
+    expected = MyBlockFactory._meta.model().clean(
+        OrderedDict(
+            [
+                ("title", "my title"),
+                ("item", OrderedDict([("label", "my-label"), ("value", 20)])),
+                ("items", []),
+                ("image", None),
+            ]
+        )
     )
+    computed.pop("items")
+    expected.pop("items")
+    assert computed == expected
 
 
 @pytest.mark.django_db
@@ -115,24 +153,34 @@ def test_custom_page_streamfield_data_complex():
     document = Document.objects.first()
 
     assert page.body[0].block_type == "char_array"
-    assert page.body[0].value == ["foo", "bar"]
+    assert page.body[2].block_type == "int_array"
+    if wagtail_version >= (2, 16):
+        assert [x.value for x in page.body[0].value.bound_blocks] == ["foo", "bar"]
+        assert [x.value for x in page.body[2].value.bound_blocks] == [100]
+    else:
+        assert page.body[0].value == ["foo", "bar"]
+        assert page.body[2].value == [100]
 
     assert page.body[1].block_type == "struct"
-    assert page.body[1].value == StructValue(
-        None,
-        [
-            ("title", "My Title"),
-            (
-                "item",
-                StructValue(None, [("label", "my-label"), ("value", 100)]),
-            ),
-            ("items", []),
-            ("image", None),
-        ],
+    computed_struct = page.body[1].value
+    expected_struct = MyBlockFactory._meta.model().clean(
+        StructValue(
+            None,
+            [
+                ("title", "My Title"),
+                (
+                    "item",
+                    StructValue(None, [("label", "my-label"), ("value", 100)]),
+                ),
+                ("items", []),
+                ("image", None),
+            ],
+        )
     )
-
-    assert page.body[2].block_type == "int_array"
-    assert page.body[2].value == [100]
+    assert eq_list_block_values(
+        computed_struct.pop("items"), expected_struct.pop("items")
+    )
+    assert computed_struct == expected_struct
 
     assert page.body[3].block_type == "image"
     assert page.body[3].value == image

--- a/tests/test_factories.py
+++ b/tests/test_factories.py
@@ -1,12 +1,12 @@
 import pytest
+from wagtail import VERSION as wagtail_version
 
 import wagtail_factories
 from tests.testapp.factories import MyTestPageFactory, MyTestPageGetOrCreateFactory
 
-try:
+if wagtail_version >= (3, 0):
     from wagtail.models import Page, Site
-except ImportError:
-    # Wagtail<3.0
+else:
     from wagtail.core.models import Page, Site
 
 
@@ -93,6 +93,10 @@ def test_custom_page_streamfield():
     assert len(page.body) == 0
 
 
+@pytest.mark.skipif(
+    wagtail_version >= (4, 1),
+    reason="Cannot assign a naive list to a ListBlock value if using ReferenceIndex",
+)
 @pytest.mark.django_db
 def test_custom_page_streamfield_data():
     root_page = wagtail_factories.PageFactory(parent=None)

--- a/tests/testapp/models.py
+++ b/tests/testapp/models.py
@@ -29,7 +29,7 @@ class MyBlock(blocks.StructBlock):
     title = blocks.CharBlock(max_length=100)
     item = MyBlockItem()
     items = blocks.ListBlock(MyBlockItem)
-    image = ImageChooserBlock()
+    image = ImageChooserBlock(required=False)
 
 
 class SimpleStructBlock(blocks.StructBlock):

